File: app/src/main/java/de/devmil/muzei/bingimageofthedayartsource/cache/BingImageCache.java
/*
 * Copyright 2014 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.muzei.bingimageofthedayartsource.cache;

import android.content.Context;
import android.net.Uri;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import de.devmil.muzei.bingimageofthedayartsource.utils.FileUtils;
import de.devmil.common.utils.LogUtil;

import static de.devmil.common.utils.LogUtil.LOGD;

/**
 * Created by devmil on 25.02.14.
 *
 * This class represents the image cache. It controls the (serialized) metadata
 * and triggers the cleanup / image download depending on the metadata it gets
 */
public class BingImageCache implements ImageDownloadRequest.OnDownloadFinishedListener {

    public static class CacheEntry
    {
        private Uri _Uri;
        private String _Description;
        private String _Copyright;

        private CacheEntry()
        {

        }

        public CacheEntry(Uri uri, String description, String copyright)
        {
            _Uri = uri;
            _Description = description;
            _Copyright = copyright;
        }

        public Uri getUri()
        {
            return _Uri;
        }

        public String getDescription()
        {
            return _Description;
        }

        public String getCopyright()
        {
            return _Copyright;
        }

        public JSONObject toJSONObject() throws JSONException {
            JSONObject result = new JSONObject();
            result.put("uri", _Uri.toString());
            result.put("desc", _Description);
            result.put("copyright", _Copyright);
            return result;
        }

        private void readFromJSONObject(JSONObject json) throws JSONException {
            _Uri = Uri.parse(json.getString("uri"));
            _Description = json.getString("desc");
            _Copyright = json.getString("copyright");
        }

        public static CacheEntry fromJSONObject(JSONObject json) throws JSONException {
            CacheEntry entry = new CacheEntry();
            entry.readFromJSONObject(json);
            return entry;
        }
    }
    public static class CacheMetadata
    {
        private CacheMetadata()
        {
        }

        public CacheMetadata(CacheEntry[] entries, Calendar validThru)
        {
            _CacheEntries = entries;
            _ValidThru = validThru;
        }

        private CacheEntry[] _CacheEntries;
        private Calendar _ValidThru;

        public CacheEntry[] getEntries()
        {
            return _CacheEntries;
        }

        public Calendar getValidThru() {
            return _ValidThru;
        }

        public void setValidThru(Calendar validThru) {
            this._ValidThru = validThru;
        }

        public JSONObject toJSONObject() throws JSONException {
            JSONObject result = new JSONObject();
            JSONArray entriesArray = new JSONArray();
            for(int i=0; i< _CacheEntries.length; i++)
            {
                entriesArray.put(_CacheEntries[i].toJSONObject());
            }
            result.put("entries", entriesArray);
            result.put("validThru", _ValidThru.getTimeInMillis());
            return result;
        }

        private void readFromJSONObject(JSONObject json) throws JSONException {
            List<CacheEntry> entries = new ArrayList<CacheEntry>();
            JSONArray entriesArray = json.getJSONArray("entries");
            for(int i=0; i<entriesArray.length(); i++)
                entries.add(CacheEntry.fromJSONObject(entriesArray.getJSONObject(i)));
            _CacheEntries = entries.toArray(new CacheEntry[0]);
            _ValidThru = Calendar.getInstance();
            _ValidThru.setTimeInMillis(json.getLong("validThru"));
        }

        public static CacheMetadata fromJSONObject(JSONObject json) throws JSONException {
            CacheMetadata result = new CacheMetadata();
            result.readFromJSONObject(json);
            return result;
        }
    }

    private static final String TAG = LogUtil.makeLogTag(BingImageCache.class);
    private static final String METADATA_FILE_NAME = "metadata.json";

    private CacheMetadata _CacheMetadata;
    private WeakReference<Context> _Context;

    private ImageDownloader _ImageDownloader;

    public BingImageCache(Context context)
    {
        _Context = new WeakReference<Context>(context);
        _ImageDownloader = new ImageDownloader(context, this);
        loadMetadata();
    }

    public void setMetadata(CacheMetadata cacheMetadata)
    {
        _CacheMetadata = cacheMetadata;
        saveMetadata();
        cleanUpFiles();
    }

    public void ensureMissingImages()
    {
        LOGD(TAG, "ensuring missing items");
        List<Uri> missingImageUris = getMissingImageUris();
        if(!missingImageUris.isEmpty())
        {
            LOGD(TAG, "start downloading missing items");
            _ImageDownloader.download(missingImageUris);
        }
        else
            LOGD(TAG, "we have all needed items => nothing to download");
    }

    private List<Uri> getMissingImageUris()
    {
        List<Uri> missingUris = new ArrayList<Uri>();
        for(CacheEntry entry : _CacheMetadata.getEntries())
        {
            if(!hasImage(entry.getUri()))
                missingUris.add(entry.getUri());
        }
        return missingUris;
    }


    public CacheMetadata getMetadata()
    {
        return _CacheMetadata;
    }

    private void saveMetadata()
    {
        File metadataFile = new File(getCacheDirectory(), METADATA_FILE_NAME);
        if(_CacheMetadata == null && metadataFile.exists())
            metadataFile.delete();
        if(_CacheMetadata == null)
            return;
        try {
            FileUtils.writeTextFile(metadataFile, _CacheMetadata.toJSONObject().toString());
        } catch (IOException e) {
        } catch (JSONException e) {
        }
    }

    /**
     * cleans up all files that are no longer needed
     * (all files that are no longer referenced by the cache)
     */
    private void cleanUpFiles()
    {
        if(_CacheMetadata == null)
            return;
        List<String> requiredFiles = new ArrayList<String>();
        for(CacheEntry entry : _CacheMetadata.getEntries())
        {
            String fName = getFileNameFromUri(entry.getUri());
            if(!requiredFiles.contains(fName))
                requiredFiles.add(fName);
        }
        requiredFiles.add(METADATA_FILE_NAME);
        for(String fName : getCacheDirectory().list())
        {
            if(!requiredFiles.contains(fName)) {
                try {
                    new File(getCacheDirectory(), fName).delete();
                }
                catch(Exception e)
                {}
            }
        }
    }

    /**
     * loads a serialized copy of the metadata from the storage
     */
    private void loadMetadata() {
        _CacheMetadata = null;
        File metadataFile = new File(getCacheDirectory(), METADATA_FILE_NAME);
        if(metadataFile.exists())
        {
            try {
                String jsonString = FileUtils.readTextFile(metadataFile);
                JSONObject jsonObject = (JSONObject)new JSONTokener(jsonString).nextValue();
                _CacheMetadata = CacheMetadata.fromJSONObject(jsonObject);
            } catch (JSONException e) {
            } catch (IOException e) {
            }
        }
    }

    private File getCacheDirectory()
    {
        Context context = _Context.get();
        if(context == null)
            return null;

        return CacheUtils.getCacheDirectory(context);
    }

    /**
     * gets called when the download of an image is finished.
     * The image gets written to the storage
     * @param content
     * @param tag
     */
    @Override
    public synchronized void downloadFinished(byte[] content, Object tag) {
        if(content != null)
        {
            Uri uri = (Uri)tag;
            String fName = getFileNameFromUri(uri);
            File f = new File(getCacheDirectory(), fName);
            try {
                if(f.exists())
                    f.delete();
                FileOutputStream fOut = new FileOutputStream(f);
                fOut.write(content);
                fOut.flush();
                fOut.close();
            } catch (FileNotFoundException e) {
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public synchronized boolean hasImage(int daysInThePast)
    {
        if(_CacheMetadata.getEntries().length <= daysInThePast)
            return false;
        return hasImage(_CacheMetadata.getEntries()[daysInThePast].getUri());
    }

    public synchronized boolean hasImage(Uri uri)
    {
        String fName = getFileNameFromUri(uri);
        return new File(getCacheDirectory(), fName).exists();
    }

    public synchronized String getFileName(int daysInThePast)
    {
        if(_CacheMetadata.getEntries().length <= daysInThePast)
            return null;
        return getFileNameFromUri(_CacheMetadata.getEntries()[daysInThePast].getUri());
    }

    final Character[] ReservedChars = {'|', '\\', '?', '*', '<', '\"', '\'', '/', ':', '>'};

    private String getFileNameFromUri(Uri uri)
    {
        String fName = uri.toString();

        for(Character c : ReservedChars)
        {
            fName = fName.replace(c, '_');
        }
        return fName;
    }
}

File: app/src/main/java/de/devmil/muzei/bingimageofthedayartsource/cache/BingImageCache.kt
/*
 * Copyright 2014 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.muzei.bingimageofthedayartsource.cache

import android.content.Context
import android.net.Uri

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.json.JSONTokener

import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.ArrayList
import java.util.Calendar

import de.devmil.muzei.bingimageofthedayartsource.utils.FileUtils
import de.devmil.common.utils.LogUtil

/**
 * Created by devmil on 25.02.14.

 * This class represents the image cache. It controls the (serialized) metadata
 * and triggers the cleanup / image download depending on the metadata it gets
 */
class BingImageCache(context: Context) : ImageDownloadRequest.OnDownloadFinishedListener {

    class CacheEntry {
        var uri: Uri? = null
            private set
        var description: String? = null
            private set
        var copyright: String? = null
            private set

        private constructor()

        constructor(uri: Uri, description: String, copyright: String) {
            this.uri = uri
            this.description = description
            this.copyright = copyright
        }

        @Throws(JSONException::class)
        fun toJSONObject(): JSONObject {
            val result = JSONObject()
            result.put("uri", uri!!.toString())
            result.put("desc", description)
            result.put("copyright", copyright)
            return result
        }

        @Throws(JSONException::class)
        private fun readFromJSONObject(json: JSONObject) {
            uri = Uri.parse(json.getString("uri"))
            description = json.getString("desc")
            copyright = json.getString("copyright")
        }

        companion object {

            @Throws(JSONException::class)
            fun fromJSONObject(json: JSONObject): CacheEntry {
                val entry = CacheEntry()
                entry.readFromJSONObject(json)
                return entry
            }
        }
    }

    class CacheMetadata {
        private constructor()

        constructor(entries: Array<CacheEntry>, validThru: Calendar) {
            this.entries = entries
            this.validThru = validThru
        }

        var entries: Array<CacheEntry>? = null
            private set
        var validThru: Calendar? = null

        @Throws(JSONException::class)
        fun toJSONObject(): JSONObject {
            val result = JSONObject()
            val entriesArray = JSONArray()
            for (i in entries!!.indices) {
                entriesArray.put(entries!![i].toJSONObject())
            }
            result.put("entries", entriesArray)
            result.put("validThru", validThru!!.timeInMillis)
            return result
        }

        @Throws(JSONException::class)
        private fun readFromJSONObject(json: JSONObject) {
            val entriesArray = json.getJSONArray("entries")
            val entries = (0..entriesArray.length() - 1).map { CacheEntry.fromJSONObject(entriesArray.getJSONObject(it)) }
            this.entries = entries.toTypedArray()
            validThru = Calendar.getInstance()
            validThru!!.timeInMillis = json.getLong("validThru")
        }

        companion object {

            @Throws(JSONException::class)
            fun fromJSONObject(json: JSONObject): CacheMetadata {
                val result = CacheMetadata()
                result.readFromJSONObject(json)
                return result
            }
        }
    }

    private var _CacheMetadata: CacheMetadata? = null
    private val _Context: WeakReference<Context> = WeakReference(context)

    private val _ImageDownloader: ImageDownloader = ImageDownloader(context, this)

    init {
        loadMetadata()
    }

    fun ensureMissingImages() {
        LogUtil.LOGD(TAG, "ensuring missing items")
        val missingImageUris = missingImageUris
        if (!missingImageUris.isEmpty()) {
            LogUtil.LOGD(TAG, "start downloading missing items")
            _ImageDownloader.download(missingImageUris)
        } else
            LogUtil.LOGD(TAG, "we have all needed items => nothing to download")
    }

    private val missingImageUris: List<Uri>
        get() {
            val missingUris = _CacheMetadata!!.entries!!
                    .filterNot { hasImage(it.uri!!) }
                    .map { it.uri!! }
            return missingUris
        }


    var metadata: CacheMetadata?
        get() = _CacheMetadata
        set(cacheMetadata) {
            _CacheMetadata = cacheMetadata
            saveMetadata()
            cleanUpFiles()
        }

    private fun saveMetadata() {
        val metadataFile = File(cacheDirectory, METADATA_FILE_NAME)
        if (_CacheMetadata == null && metadataFile.exists())
            metadataFile.delete()
        if (_CacheMetadata == null)
            return
        try {
            FileUtils.writeTextFile(metadataFile, _CacheMetadata!!.toJSONObject().toString())
        } catch (e: IOException) {
        } catch (e: JSONException) {
        }

    }

    /**
     * cleans up all files that are no longer needed
     * (all files that are no longer referenced by the cache)
     */
    private fun cleanUpFiles() {
        if (_CacheMetadata == null)
            return
        val requiredFiles = ArrayList<String>()
        _CacheMetadata!!.entries!!
                .map { getFileNameFromUri(it.uri!!) }
                .filterNot { requiredFiles.contains(it) }
                .forEach { requiredFiles.add(it) }
        requiredFiles.add(METADATA_FILE_NAME)
        cacheDirectory!!.list()
                .filterNot { requiredFiles.contains(it) }
                .forEach {
                    try {
                        File(cacheDirectory, it).delete()
                    } catch (e: Exception) {
                    }
                }
    }

    /**
     * loads a serialized copy of the metadata from the storage
     */
    private fun loadMetadata() {
        _CacheMetadata = null
        val metadataFile = File(cacheDirectory, METADATA_FILE_NAME)
        if (metadataFile.exists()) {
            try {
                val jsonString = FileUtils.readTextFile(metadataFile)
                val jsonObject = JSONTokener(jsonString).nextValue() as JSONObject
                _CacheMetadata = CacheMetadata.fromJSONObject(jsonObject)
            } catch (e: JSONException) {
            } catch (e: IOException) {
            }

        }
    }

    private val cacheDirectory: File?
        get() {
            val context = _Context.get() ?: return null

            return CacheUtils.getCacheDirectory(context)
        }

    /**
     * gets called when the download of an image is finished.
     * The image gets written to the storage
     * @param content
     * *
     * @param tag
     */
    @Synchronized override fun downloadFinished(content: ByteArray?, tag: Any) {
        if (content != null) {
            val uri = tag as Uri
            val fName = getFileNameFromUri(uri)
            val f = File(cacheDirectory, fName)
            try {
                if (f.exists())
                    f.delete()
                val fOut = FileOutputStream(f)
                fOut.write(content)
                fOut.flush()
                fOut.close()
            } catch (e: FileNotFoundException) {
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    @Synchronized fun hasImage(daysInThePast: Int): Boolean {
        if (_CacheMetadata!!.entries!!.size <= daysInThePast)
            return false
        return hasImage(_CacheMetadata!!.entries!![daysInThePast].uri!!)
    }

    @Synchronized fun hasImage(uri: Uri): Boolean {
        val fName = getFileNameFromUri(uri)
        return File(cacheDirectory, fName).exists()
    }

    @Synchronized fun getFileName(daysInThePast: Int): String? {
        if (_CacheMetadata!!.entries!!.size <= daysInThePast)
            return null
        return getFileNameFromUri(_CacheMetadata!!.entries!![daysInThePast].uri!!)
    }

    internal val ReservedChars = arrayOf('|', '\\', '?', '*', '<', '\"', '\'', '/', ':', '>')

    private fun getFileNameFromUri(uri: Uri): String {
        var fName = uri.toString()

        for (c in ReservedChars) {
            fName = fName.replace(c, '_')
        }
        return fName
    }

    companion object {

        private val TAG = LogUtil.makeLogTag(BingImageCache::class.java)
        private val METADATA_FILE_NAME = "metadata.json"
    }
}
